docs/development/writingzeppelinspell.md

```diff
@@ -0,0 +1,213 @@
+
+### 2. Using
+
+Spell works like an interpreter. Thus, use the `MAGIC` to execute spell in a note. (you might need to refresh after enabling)
+
+<img class="img-responsive" style="width:70%" src="../assets/themes/zeppelin/img/docs-img/writing_spell_using.gif" />
+
+
+## Write a new Spell 
+
+Making a new spell is similar to [Helium Visualization#write-new-visualization](./writingzeppelinvisualization.html#write-new-visualization).
+
+- Add framework dependency called zeppelin-spell into `package.json`
+- Write code using framework
+- Publish your spell to [npm]("https://www.npmjs.com/")
+
+### 1. Create a npm package
+
+Create a [package.json](https://docs.npmjs.com/files/package.json) in new directory for spell. 
+
+- You have to add a framework called `zeppelin-spell` as a dependency to create spell ([zeppelin-spell](https://github.com/apache/zeppelin/tree/master/zeppelin-web/src/app/spell))
+- Also, you can add any dependencies you want to utilise.
+
+Here's an example
+
+```
+{
+  "name": "zeppelin-echo-spell",
+  "description": "Zeppelin Echo Spell (example)",
+  "version": "1.0.0",
+  "main": "index",
+  "author": "",
+  "license": "Apache-2.0",
+  "dependencies": {
+    "zeppelin-spell": "*"
+  },
+  "helium": {
+    "icon" : "<i class='fa fa-repeat'></i>",
+    "spell": {
+      "magic": "%echo",
+      "usage": "%echo <TEXT>"
+    }
+  }
+}
+```
+
+### 2. Write spell using framework 
+
+Here are some examples you can refer
+
+- [Echo Spell](https://github.com/apache/zeppelin/blob/master/zeppelin-examples/zeppelin-example-spell-echo/index.js)
+- [Markdown Spell: Using library](https://github.com/apache/zeppelin/blob/master/zeppelin-examples/zeppelin-example-spell-markdown/index.js)
+- [Flowchart Spell: Using DOM](https://github.com/apache/zeppelin/blob/master/zeppelin-examples/zeppelin-example-spell-flowchart/index.js)
+- [Google Translation API Spell: Using API (returning promise)](https://github.com/apache/zeppelin/blob/master/zeppelin-examples/zeppelin-example-spell-translator/index.js)
+
+Now, you need to write code to create spell which processing text.
+
+```javascript
+import {
+    SpellBase,
+    SpellResult,
+    DefaultDisplayType,
+} from 'zeppelin-spell';
+
+export default class EchoSpell extends SpellBase {
+    constructor() {
+        /** pass magic to super class's constructor parameter */
+        super("%echo");
+    }
+
+    interpret(paragraphText) {
+        const processed = paragraphText + '!';
+        
+        /** 
+         * should return `SpellResult` which including `data` and `type`
+         * default type is `TEXT` if you don't specify.  
+         */
+        return new SpellResult(processed);
+    }
+}
+```
+
+Here is another example. Let's say we want to create markdown spell. First of all, we should add a dependency for markdown in package.json
+
+```
+// package.json
+ "dependencies": {
+    "markdown": "0.5.0",
```

Comment:
@felixcheung Thanks for detailed review. 

The version value `*` means, these modules are provided by zeppelin. So it's valid and proper.

- Currently, Zeppelin doesn't publish any NPM packages. So Zeppelin provides them by copying not downloading. 
- If we need versioning for them, we have to publish `zeppelin-spell` modules to NPM first of all.